### PR TITLE
refactor: deduplicate util function from admin/server modules

### PR DESCRIPTION
**Pull request by furlongmt**
…single utils class

Signed-off-by: Matt Furlong <furlngm@google.com>

Commit Message: Deduplicate utility function that was copy-pasted from admin module into the server module.
Risk Level: Low.
Testing: Unit test was refactored into server directory.

### CODE REVIEW

File: source/server/admin/server_info_handler.cc [hunk 1]
```diff
@@ -2,9 +2,10 @@
 
 #include "envoy/admin/v3/memory.pb.h"
 
+#include "source/common/http/headers.h"
```
Comment (abeyad) [inline source/server/admin/server_info_handler.cc:5]: why did this header get added?  may have been VS Code automatically adding it

File: source/server/server.cc [hunk 2]
```diff
@@ -60,6 +60,7 @@
 #include "source/server/listener_manager_factory.h"
 #include "source/server/regex_engine.h"
 #include "source/server/ssl_context_manager.h"
+#include "source/server/utils.h"
```
Comment (abeyad) [inline source/server/server.cc:63]: you'll also want to remove the serverState function in the anonymous namespace of this file

File: source/server/utils.cc [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+    return health_check_failed ? envoy::admin::v3::ServerInfo::DRAINING
+                               : envoy::admin::v3::ServerInfo::LIVE;
+  }
+  IS_ENVOY_BUG("unexpected server state enum");
+  return envoy::admin::v3::ServerInfo::PRE_INITIALIZING;
+}
+
+} // namespace Utility
+} // namespace Server
+} // namespace Envoy
```
Comment (abeyad) [inline source/server/utils.cc:26]: you need a newline at the end of the file, or the formatters complain

File: source/server/utils.h [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+namespace Envoy {
+namespace Server {
+namespace Utility {
+
+envoy::admin::v3::ServerInfo::State serverState(Init::Manager::State state,
+                                                bool health_check_failed);
+
+} // namespace Utility
+} // namespace Server
+} // namespace Envoy
```
Comment (abeyad) [inline source/server/utils.h:15]: you need a newline at the end of the file, or the formatters complain

File: test/server/BUILD [hunk 5]
```diff
@@ -378,3 +378,11 @@ envoy_benchmark_test(
...
     benchmark_binary = "server_stats_flush_benchmark",
 )
+
+envoy_cc_test(
+    name = "utils_test",
+    srcs = envoy_select_admin_functionality(["utils_test.cc"]),
+    deps = [
+        "//source/server:utils_lib",
+    ],
+)
```
Comment (abeyad) [inline test/server/BUILD:388]: you need a newline at the end of the file, or the formatters complain

File: test/server/utils_test.cc [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+// Most utils paths are covered through other tests, these tests take of
+// of special cases to get remaining coverage.
+TEST(UtilsTest, BadServerState) {
+  Utility::serverState(Init::Manager::State::Uninitialized, true);
+  EXPECT_ENVOY_BUG(Utility::serverState(static_cast<Init::Manager::State>(123), true),
+                   "unexpected server state");
+}
+} // namespace Utilities
+} // namespace Server
+} // namespace Envoy
```
Comment (abeyad) [inline test/server/utils_test.cc:20]: you need a newline at the end of the file, or the formatters complain

File: test/server/utils_test.cc [hunk 7]
```diff
@@ -0,0 +1,20 @@
+#include "source/server/utils.h"
+
+#include "test/test_common/utility.h"
+
+#include "gtest/gtest.h"
+
+namespace Envoy {
+namespace Server {
+namespace Utilities {
```
Comment (abeyad) [inline test/server/utils_test.cc:9]: this namespace is plural whereas elsewhere you named it just `Utility`... probably want to unify these

**Review by abeyad — COMMENTED** [review]
Awesome, thanks Matt!

File: source/server/server.cc (hunk 2 above)
Comment (furlongmt) [reply source/server/server.cc:63]: Ah whoops, I meant to do that before pushing.

File: test/server/utils_test.cc (hunk 7 above)
Comment (furlongmt) [reply test/server/utils_test.cc:9]: Makes sense. Updated.

File: source/server/admin/server_info_handler.cc (hunk 1 above)
Comment (furlongmt) [reply source/server/admin/server_info_handler.cc:5]: I think it was being pulled in as a transitive dependency before. Without this header, the build fails:
```
source/server/admin/server_info_handler.cc: In member function 'Envoy::Http::Code Envoy::Server::ServerInfoHandler::handlerMemory(Envoy::Http::ResponseHeaderMap&, Envoy::Buffer::Instance&, Envoy::Server::AdminStream&)':
source/server/admin/server_info_handler.cc:45:50: error: 'Envoy::Http::Headers' has not been declared
   45 |   response_headers.setReferenceContentType(Http::Headers::get().ContentTypeValues.Json);
```

**Review by abeyad — APPROVED** [review]
Thanks Matt, looks good!  Let's see what Alyssa says about the location of the utils code.

**Comment by abeyad** [comment]
@furlongmt the presubmit is legit, the fix is in the output: https://dev.azure.com/cncf/envoy/_build/results?buildId=123048&view=logs&j=55bc2ab9-36c1-5ab8-58a6-8c258728eeff&t=33b57bf9-ef13-58c8-bc2b-f804aeb16d39&l=606

**Comment by abeyad** [comment]
Adding @lizan for cross-company review

**Comment by abeyad** [comment]
@furlongmt The presubmit failures are legitimate (one of them was that the namespace wasn't specified, i.e. `Utility::serverState` instead of just `serverState`)

**Review by alyssawilk — COMMENTED** [review]
Looks good! 
Can you update your PR description and fill in the relevant fields and remove the irrelevant ones?

/wait

**Review by lizan — APPROVED** [review]

**Pull request merged**